622 Add conformance rules explanation (#1447)
AbsaOSS/enceladus: _docs/2.0.0/usage/menas-conformance-rules.md

---
layout: docs
title: Usage - Menas Conformance Rules
version: '2.0.0'
categories:
    - '2.0.0'
    - usage
---

## Table Of Contents
<!-- toc -->
- [Table Of Contents](#table-of-contents)
- [Intro](#intro)
- [Casting Conformance Rule](#casting-conformance-rule)

> - [Concatenation Conformance Rule](#concatenation-conformance-rule)

Reviewer: I think that the Coalesce Conformance rule should be added.

Author: Done

Reviewer 2: I didn't want to bother Sasa with it. Planned to add it after merge. Thanks Sasa

Author: I cheated. I used what you wrote in Release Notes 🚀

- [Drop Conformance Rule](#drop-conformance-rule)
- [Literal Conformance Rule](#literal-conformance-rule)
- [Mapping Conformance Rule](#mapping-conformance-rule)
- [Negation Conformance Rule](#negation-conformance-rule)
- [SingleColumn Conformance Rule](#singlecolumn-conformance-rule)
- [SparkSessionConf Conformance Rule](#sparksessionconf-conformance-rule)
- [Uppercase Conformance Rule](#uppercase-conformance-rule)
- [FillNulls Conformance Rule](#fillnulls-conformance-rule)
<!-- tocstop -->

## Intro

Conformance Rules are a way for the user to enhance the data. Every conformance rule has an output column and an option for running a control measure. Output column defines the name of the output column into which the rule will output. Control measure check will run an Atum control measure check as in previouse stages as defined in controls of `_INFO` file.

Every column also has one or more input columns.

**Important** - We never override a column. Each rule produces a new column.

## Casting Conformance Rule

Allows users to cast a specific column to another type. This conformance rules provides a selection of other types.

> 

Reviewer: I think it would be nice to add some specifications. 
Like allowed "conversions"
- any type to string is ok 
- boolean to number is ok (false -> 0; true -> 1)
- number to boolean is ok  (0 -> false; any -> true)
- to binary type is only from string type

## Concatenation Conformance Rule

> Concatenation rule concatenates multiple input columns together into a single column. Columns are frist transformed into their string representation.

Reviewer 3: ```suggestion
Concatenation rule concatenates multiple input columns together into a single column. Columns are first transformed into their string representation.
```

## Drop Conformance Rule

> Drop Conformance rule remove a column from the DataFrame.

Reviewer 3: Maybe just Remove a column from the DataFrame.

Author: You mean omit the name?

Reviewer 3: Yes, seems like repeating

Author: It is consistent with every other entry. I find it easier to read a full sentence

Reviewer 4: Would it sound better if we say 'Dataset" instead of "dataframe"

Author: Somewhere else I. removed it to output data.

## Literal Conformance Rule

Literal rule add a column with a string literal provided by the user.

## Mapping Conformance Rule

> To use a mapping conformance rule, user first needs to define a Mapping Table in menas. Mapping Tables have the same properties and rules around them as Datasets, plus it is expected of them to be in a parquet format.

Reviewer 4: I would remove "plus"

> To use a mapping conformance rule, user first needs to define a Mapping Table in menas. Mapping Tables have the same properties and rules around them as Datasets, plus it is expected of them to be in a parquet format.

Author: Removed `plus`, added `and`.

> When defining a mapping conformance rule, users first needs to pick a correct Mapping Table and correct version. Then there is a question if the data can have Null values in join conditions. This means if the join that will be executed should be null safe or not.

Reviewer 4: line 55. would it sound better if it reads " If a user requires the data to match NUll values in  join condition (s) , the join will happen even if both the table and .....null."

> When defining a mapping conformance rule, users first needs to pick a correct Mapping Table and correct version. Then there is a question if the data can have Null values in join conditions. This means if the join that will be executed should be null safe or not.

Author: So, I might misunderstand this null safety, but I thought it is about the fact that one column can have nulls. Not that both keys are null.

> When defining a mapping conformance rule, users first needs to pick a correct Mapping Table and correct version. Then there is a question if the data can have Null values in join conditions. This means if the join that will be executed should be null safe or not.

Reviewer 2: How I understand it, the difference is, if `null` on both side produce a match or not.
I am not sure how would the rule ensure not being present in the key or mapping columns? 🤔

> When defining a mapping conformance rule, users first needs to pick a correct Mapping Table and correct version. Then there is a question if the data can have Null values in join conditions. This means if the join that will be executed should be null safe or not.

Author: If you have null on one side and this is not checked, then you get a conformance error.

> Then the join conditions. A join condition condition conveys a relationship between the Dataset and MappingTable. The join condition specifies how the rows from one table will be combined with the rows of another table. This is based on the equality of the rows.

Reviewer 2: ```suggestion
Following, a join condition condition conveys a relationship between the table data and Mapping Table. The join condition specifies how the rows from one table will be combined with the rows of the other table. This is based on the equality of the rows.
```

Author: - I don't like this following part because it is jumping to the next section without tellin user we are moving to the next section
- I don't like the table data. We have never spoken about table data, we always speak about Dataset

Reviewer 2: Dataset is good, I think. Would make it `dataset`, I don't see the reason to capitalize it.

Reviewer 2: I think the _Following_ signals it's another section. But what about:
```suggestion
Then the join conditions conveys the relationship between the Dataset and Mapping Table. The join condition specifies how the rows from one table will be combined with the rows of another table. This is based on the equality of the values in the selected columns.
```
Sorry, changed the end too, not the whole rows are relevant.

Last is the target column, which specifies which column from the mapping table will be written into the output column.

## Negation Conformance Rule

Negation conformance rule negates any Numerical or Boolean value.

## SingleColumn Conformance Rule

> Single column conformance rule transforms column into a column of structs of previous columns values. Input column alias here will be the name/key of the struct

Reviewer 4: I find this a little confusing as well

> Single column conformance rule transforms column into a column of structs of previous columns values. Input column alias here will be the name/key of the struct

Author: Ok, you know what it does, ideas?

> Single column conformance rule transforms column into a column of structs of previous columns values. Input column alias here will be the name/key of the struct

Reviewer 4: Takes data from Dataset columns into conformed column with new strcuct type?

> Single column conformance rule transforms column into a column of structs of previouse columns values. Input column alias here will be the name/key of the struct

Reviewer 3: ```suggestion
Single column conformance rule transforms column into a column of structs of previous columns values. Input column alias here will be the name/key of the struct
```

- [Casting Conformance Rule](#casting-conformance-rule)
- [Concatenation Conformance Rule](#concatenation-conformance-rule)
- [Drop Conformance Rule](#drop-conformance-rule)
- [FillNulls Conformance Rule](#fillnulls-conformance-rule)
- [Literal Conformance Rule](#literal-conformance-rule)
- [Mapping Conformance Rule](#mapping-conformance-rule)
- [Negation Conformance Rule](#negation-conformance-rule)
- [SingleColumn Conformance Rule](#singlecolumn-conformance-rule)
- [SparkSessionConf Conformance Rule](#sparksessionconf-conformance-rule)
- [Uppercase Conformance Rule](#uppercase-conformance-rule)
<!-- tocstop -->

## Intro

Conformance Rules are a way for the user to enhance the data. Every conformance rule has an output column and an option for running a control measure. Output column defines the name of the output column into which the rule will output. Control measure check will run an Atum control measure check as in previous stages as defined in controls of `_INFO` file.

Every column also has one or more input columns.

**Important** - We never override a column. Each rule produces a new column.

## Casting Conformance Rule

> Casting conformance rule Allows users to cast a specific column to another type. This conformance rule provides a selection of other types.

Reviewer 2: ```suggestion
Casting conformance rule allows users to cast a specific column to another type. This conformance rule provides a selection of other types.
```

## Concatenation Conformance Rule

Concatenation conformance rule concatenates two or more input columns together into a single column. Columns are first transformed into their string representation before the actual concatenation.

## Drop Conformance Rule

Drop conformance rule removes a column from the output data.

## FillNulls Conformance Rule

FillNulls conformance rule takes a column and replaces all the nulls with a literal provided by the user.

## Literal Conformance Rule

Literal conformance rule adds a column with a string literal provided by the user.

## Mapping Conformance Rule

To use a mapping conformance rule, the user first needs to define a Mapping Table in Menas. Mapping Tables have the same properties and rules around them as Datasets, and it is expected of them to be in a parquet format.

When defining a mapping conformance rule, users first need to pick a correct Mapping Table and correct version. Then there is a question if the data can have Null values in join conditions. This means if the join that will be executed should be null safe or not.

> Then the join conditions. A join condition conveys a relationship between the Dataset and MappingTable. The join condition specifies how the rows from one table will be combined with the rows of the other table. This is based on the equality of the rows.

Reviewer 2: **Comment:** Can't help but find the single sentence _"Then the join conditions."_ at the beginning of the paragraph not fitting. 🤓

> Then the join conditions. A join condition conveys a relationship between the Dataset and MappingTable. The join condition specifies how the rows from one table will be combined with the rows of the other table. This is based on the equality of the rows.

Reviewer 2: **Comment:**
_This is based on the equality of the rows._
I have doubts of correctness of this statement. I think it might confuse users. 🤔

Author: Ok. I applied your previous comment. You have 3 on this one line. I am not sure what to do.
